Install bot2-procman and bot-spy

fbudin69500 opened the pull request:
Needed for https://github.com/RobotLocomotion/spartan/pull/116. FYI @manuelli.

<!-- Reviewable:start -->
----
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7068)
<!-- Reviewable:end -->

jamiesnape commented:
Reviewed 2 of 6 files at r1, 2 of 9 files at r2.
Review status: 4 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZQep0yOr0zCGMUSRb:-KuZQep0yOr0zCGMUSRc:b-axd0kd) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L6)):*
> ```
> readonly prefix=$(python -c 'import os; print(os.path.realpath(os.path.join("'"$0"'", os.pardir, os.pardir)))')
> 
> export PYTHONPATH=${prefix}/lib/python2.7/dist-packages:${prefix}/lib/python2.7/site-packages:${PYTHONPATH}
> ```

FYI On Mac,  everything needed by this goes into `site-packages`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Reviewed 1 of 6 files at r1, 2 of 9 files at r2.
Review status: 7 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
So I can review this under the assumption that it's a thing we should do, but I'm curious why?  What would prevent Spartan from building `RobotLocomotion/libbot2` on its own, using its fine pre-existing CMake build system?  In fact, Drake only uses a few Java bits of libbot (in contrast to C++ code, where the compiled ABI matters).  So why is Drake installing libbot2 _at all_?  In other words, why can't it be a private dependency of Drake?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuZR0Ebp1Cx0n2D1P2J:b-lvws4r)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Not just Spartan, Director as well. The specification was that if we used an external in Drake, then it should be available if if it were built by CMake.

---

Review status: 7 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuZSCIqYNFES9V3lAsl:bvxfhy4)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
(as if)

---

Review status: 7 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuZSblLN5mvrfaYnCyw:bv2fkg1)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Hmm.  If that's really the rules of engagement (but why? private dependencies are a fine thing), then IMO we should just remove libbot2 from Drake entirely.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuZSb0ufxRioEpwX3wM:bl7t41c)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
I have no objections either way. I guess it was just keeping the status quo.

---

Review status: 7 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuZT9_DRwc3n5Nm6SvT:bj8ec1q)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 6 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZQep0yOr0zCGMUSRb:-KuZWsIbvFF2wCcIu-sP:b4imb7d) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L6)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

FYI On Mac,  everything needed by this goes into `site-packages`. 
</blockquote></details>

On xenial too actually. I simplified it. The only packages installed in dist-packages are director and urdf_parser_py .

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 6 files at r1, 6 of 9 files at r2.
Review status: 7 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[tools/drake.bzl, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZWfhWVGZm3bjSgsUr:-KuZWfhWVGZm3bjSgsUs:b-4niefw) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L126)):*
> ```Python
>         target = ":"+name,
>         filename = launcher_name + ".sh",
>         )
> ```

Should (and could) the `-launcher` have a different visibility declared?  Such that only the install system can see it?

---

*[tools/install.py.in, line 101 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZVpZnQzSZqC3Fc5t2:-KuZVpZonSc_e9RJ4bZD:bgrzhq2) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.py.in#L101)):*
> ```
> 
> def create_java_launcher(filename, classpath, main_class):
>     # Make sure install directory exists
> ```

BTW punc

---

*[tools/install.py.in, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZVut4fxbZzF6Anpfs:-KuZVut4fxbZzF6Anpft:bgrzhq2) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.py.in#L106)):*
> ```
>     if not os.path.exists(filepath):
>         os.makedirs(filepath)
>     # Write launcher
> ```

BTW punc

---

*[tools/libbot.BUILD, line 582 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZVFalcZIlNAjOuLfr:-KuZVFalcZIlNAjOuLfs:b-93irfr) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/libbot.BUILD#L582)):*
> ```
>     hdr_dest = HDR_DEST,
>     hdr_strip_prefix = ["bot2-core/src"],
>     deps = ["@lcm//:install", "@lcmtypes_bot2_core//:install"],  # rename
> ```

Unclear what "rename" means here.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZTMp_id4Z4t5DrLAf:-KuZTMpaW7UKDwr4Xy3r:bm7im4g) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L1)):*
> ```
> #!/bin/sh
> ```

`tools/third_party/` is reserved for file copied (or forked) from upstream.  Are these  from upstream?  I didn't immediately find them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 9 files at r2.
Review status: 7 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[tools/drake.bzl, line 95 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZtMA767S3b7bEZBF3:-KuZtMA767S3b7bEZBF4:b-828qjm) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L95)):*
> ```Python
>         "main_class": attr.string(mandatory = True),
>         "filename": attr.string(),
>         "target": attr.label(),
> ```

Unclear why the second two attrs are not flagged as mandatory.

---

*[tools/drake.bzl, line 108 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZqwI9ajy8mgKVNhUD:-KuZqwIAuEryiA-swpLv:bc1el8h) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L108)):*
> ```Python
>     **kwargs):
>     """Creates a rule to declare a java binary and a MainClassInfo Provider
>     
> ```

BTW trailing whitespace

---

*[tools/drake.bzl, line 124 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZr5plgBropmXgHn1K:-KuZr5plgBropmXgHn1L:b-opeylc) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L124)):*
> ```Python
>         name = launcher_name,
>         main_class = main_class,
>         target = ":"+name,
> ```

BTW the lack of whitespace around `+` is inconsistent with nearby lines in this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Finished first pass.

---

Reviewed 1 of 9 files at r2.
Review status: 7 of 11 files reviewed at latest revision, 17 unresolved discussions.

---

*[tools/install.bzl, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZtdzColaYvjlV-3eK:-KuZtdzColaYvjlV-3eL:bg6mecy) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L4)):*
> ```Python
> 
> load("@drake//tools:pathutils.bzl", "dirname", "output_path", "join_paths")
> load("@drake//tools:drake.bzl","MainClassInfo")
> ```

BTW missing space after `,`.

---

*[tools/install.bzl, line 210 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZuXmM-Lv7zWyZoG_R:-KuZuXmM-Lv7zWyZoG_S:bbc7x7z) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L210)):*
> ```Python
>     excluded_files = []
>     if target.files_to_run.executable:
>         excluded_files = [_output_path(ctx, target.files_to_run.executable, warn_foreign = False)]
> ```

FYI Here (and many places below), try to stay <= 79 chars when possible.

---

*[tools/install.bzl, line 234 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1wHAb-3HsxY3S239:-Ku_1wHAb-3HsxY3S23A:bgrzhq2) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L234)):*
> ```Python
> def _install_java_launcher_actions(ctx, dest, java_dest, java_strip_prefix, rename, target):
>     main_class = target[MainClassInfo].main_class
>     # List runtime_classpath and compute their install paths
> ```

BTW punc

---

*[tools/install.bzl, line 240 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_27_5jUdpEvTXlJUf:-Ku_27_5jUdpEvTXlJUg:b-auhylh) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L240)):*
> ```Python
>     for jar in target[MainClassInfo].classpath:
>         jar_install = _install_action(ctx, jar, java_dest, java_strip_prefix, rename )
>         classpath += "\"" + join_paths("$prefix", jar_install.dst) + "\" "
> ```

Unclear why the value is double-quoted here.  I would have expected the receiving code (which I think is _java_launcher_code) to handle the quoting, and in particular for the `repr` there to already be quoting for us.

---

*[tools/install.bzl, line 246 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1zTU-NiSeyLZnwCf:-Ku_1zTVzRBhD_NdMNnc:bsb4207) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L246)):*
> ```Python
>     file_dest = join_paths(dest, filename)
>     file_dest = _rename(file_dest, rename)
>     
> ```

BTW trailing whitespace

(I would have thought that bzlcodestyle would flag these.  I guess not.)

---

*[tools/install.bzl, line 305 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1kynL6VEftkzi4gk:-Ku_1kynL6VEftkzi4gl:bhwcqc6) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L305)):*
> ```Python
>                            )
>         else:
>             actions += _install_runtime_actions(ctx, t)
> ```

It's not clear to me why having not matched any of the "if-else" clauses above, that we know we have a "runtime action".  What is a runtime action, and why is it the base case?

---

*[tools/install.bzl, line 312 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZvSr8eUWqNS0KB01z:-KuZvSr9yPXGb0Qmmuog:b-ze7mau) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L312)):*
> ```Python
>     for a in actions:
>         if a.dst not in installed_files:
>             if not hasattr(a, "main_class"):
> ```

This probably deserves a comment about where `"main_class"`) is coming from -- is that the java rule's attribute, or the MainClassInfo field?

Or capture it in a new predicate function above like `is_java_launcher` and use that here and in the list comprehension filter ~20 lines below also?

---

*[tools/install.bzl, line 426 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZv-UKwJMJ-cTrwL09:-KuZv-ULAp-QfIKywJip:b-94lbwd) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L426)):*
> ```Python
>     including e.g. shared libraries — in a target's ``runfiles``, only *source*
>     artifacts are considered when guessing resource files.
>     
> ```

BTW training whitespace

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-Ku_2UyZXB7dS3waXHuf:b-xhgvng)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[tools/drake.bzl, line 95 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZtMA767S3b7bEZBF3:-KudfV9OysFT8ecdt5n_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L95)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why the second two attrs are not flagged as mandatory.
</blockquote></details>

Done.

---

*[tools/drake.bzl, line 108 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZqwI9ajy8mgKVNhUD:-KudfnvmNkkakkudtmgP:brsyzs1) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L108)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW trailing whitespace
</blockquote></details>

Corrected in the previous patch.

---

*[tools/drake.bzl, line 124 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZr5plgBropmXgHn1K:-KudfrBFuugG3HKY0j8z:brsyzs1) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L124)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW the lack of whitespace around `+` is inconsistent with nearby lines in this file.
</blockquote></details>

Corrected in the previous patch.

---

*[tools/drake.bzl, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZWfhWVGZm3bjSgsUr:-KudrizBeHo0ObmcSxVI:b-t56sf4) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/drake.bzl#L126)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Should (and could) the `-launcher` have a different visibility declared?  Such that only the install system can see it?
</blockquote></details>

I was unsuccessful trying to limit the visibility to the install system (it was generating a cycle in the dependency graph. But I added a visibility flag to the drake launcher rule.

---

*[tools/install.bzl, line 4 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZtdzColaYvjlV-3eK:-Kuds81Yzm9CfqaT-v2u:b-1hq4vk) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW missing space after `,`.
</blockquote></details>

Corrected in previous patch.

---

*[tools/install.bzl, line 210 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZuXmM-Lv7zWyZoG_R:-KudsC_uah5PeVr8NjgZ:b-1hq4vk) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L210)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Here (and many places below), try to stay <= 79 chars when possible.
</blockquote></details>

Corrected in previous patch.

---

*[tools/install.bzl, line 234 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1wHAb-3HsxY3S239:-KudsM56wRcFpYYAJR8u:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L234)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW punc
</blockquote></details>

Done.

---

*[tools/install.bzl, line 240 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_27_5jUdpEvTXlJUf:-KudscLzLy5VWj_76Xdz:bkc0otk) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L240)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why the value is double-quoted here.  I would have expected the receiving code (which I think is _java_launcher_code) to handle the quoting, and in particular for the `repr` there to already be quoting for us.
</blockquote></details>

The classpath string is used to generate a shell script. Potentially (although unlikely) the value contained in `jar_install.dst` contains white spaces which will create issues when running the shell script. Comment added in the source code.

---

*[tools/install.bzl, line 246 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1zTU-NiSeyLZnwCf:-KudtLgcc6YRh4AoNqxq:be9977d) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L246)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW trailing whitespace

(I would have thought that bzlcodestyle would flag these.  I guess not.)
</blockquote></details>

removed in previous patch.

---

*[tools/install.bzl, line 305 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1kynL6VEftkzi4gk:-KudtVPcW9yVjmWpVrpY:bft7n47) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L305)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It's not clear to me why having not matched any of the "if-else" clauses above, that we know we have a "runtime action".  What is a runtime action, and why is it the base case?
</blockquote></details>

If one wants to install an executable script that is copied from the source directory, none of the above "if-else" clauses will work. Comment added in the source code.

---

*[tools/install.bzl, line 312 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZvSr8eUWqNS0KB01z:-Kue-3LH1mfp1WH_kkz5:bxnuesu) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L312)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This probably deserves a comment about where `"main_class"`) is coming from -- is that the java rule's attribute, or the MainClassInfo field?

Or capture it in a new predicate function above like `is_java_launcher` and use that here and in the list comprehension filter ~20 lines below also?
</blockquote></details>

I updated the code so that it does not rely on "main_class" anymore, but on src. That is what I wanted to do initially, but I hadn't found a way of coding that that was satisfying.

---

*[tools/install.bzl, line 426 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZv-UKwJMJ-cTrwL09:-Kue-HNQOIdoe5MAJIhY:b5r9ju8) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L426)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW training whitespace
</blockquote></details>

corrected in previous patch.

---

*[tools/install.py.in, line 101 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZVpZnQzSZqC3Fc5t2:-Kue-RISfRVyPoNrqBxt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.py.in#L101)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW punc
</blockquote></details>

Done.

---

*[tools/install.py.in, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZVut4fxbZzF6Anpfs:-Kue-RzTooqazChTJkDJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.py.in#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW punc
</blockquote></details>

Done.

---

*[tools/libbot.BUILD, line 582 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZVFalcZIlNAjOuLfr:-Kue03QAjKpy-XUpeyUU:bp53pft) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/libbot.BUILD#L582)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear what "rename" means here.
</blockquote></details>

I am guessing you are referring to "# rename" that is just a few lines below. I don't know either what it is there for. I removed it.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZTMp_id4Z4t5DrLAf:-Kue0zywBO2JjvE78MTh:bpo7218) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`tools/third_party/` is reserved for file copied (or forked) from upstream.  Are these  from upstream?  I didn't immediately find them.
</blockquote></details>

https://github.com/RobotLocomotion/libbot2/blob/master/bot2-procman/python/scripts/bot-procman-sheriff

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
The push isn't showing up yet.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-Kue29u-G0oDdyEmdOz-:b5iunkf)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Sorry, I was running the style checks locally. It is pushed now.

---

Review status: 6 of 11 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-Kue2YZ6ClNc3efmtcys:bn3tttc)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Reviewed 1 of 4 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[tools/install.bzl, line 240 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_27_5jUdpEvTXlJUf:-Kue5Eivaqdm9d4xNnrS:b5ez0ot) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L240)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

The classpath string is used to generate a shell script. Potentially (although unlikely) the value contained in `jar_install.dst` contains white spaces which will create issues when running the shell script. Comment added in the source code.
</blockquote></details>

Ah, I think I can follow it now.  I still think this is a bad-code-smell place to perform the path escaping / quoting.  It is extremely spooky action-at-a-distance (well, really invariants-maintined-at-a-distance).

It's the bash emission where the quoting matters, so it seems to me like the body of `create_java_launcher` is where the quoting should happen.  The classpath should be passed down as a list in all bzl and python (the list would be repr'd into the generated install python), and the implementation of `create_java_launcher` should join and quote it.

---

*[tools/install.bzl, line 305 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1kynL6VEftkzi4gk:-Kue3StCGWOHB8ak6z3O:ba85aw6) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L305)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

If one wants to install an executable script that is copied from the source directory, none of the above "if-else" clauses will work. Comment added in the source code.
</blockquote></details>

Sure, I understand what case we're trying to handle.  What's still unclear is why having failed all of the prior if-tests, we know for sure that we have an executable script from the source directory?  Are there truly no other kinds of targets?

I'm looking for either a comment explaining why only this one kind of target could possibly remain, or replacing the `else` guard with an if-test that positively identifies executable scripts from the source directory.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZTMp_id4Z4t5DrLAf:-Kue35twvxIJZ2g3nY9s:bctavmm) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L1)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

https://github.com/RobotLocomotion/libbot2/blob/master/bot2-procman/python/scripts/bot-procman-sheriff
</blockquote></details>

I see it's the same filename now, thanks.  But this script isn't really a fork of that, is it?  It seems more like a reimplementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZTMp_id4Z4t5DrLAf:-KueaqWVdhVPngbDvIgv:bh59esl) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I see it's the same filename now, thanks.  But this script isn't really a fork of that, is it?  It seems more like a reimplementation.
</blockquote></details>

It is indeed a reimplementation. Should I move it directly into `tools`? That doesn't seem to be the right place either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 9 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[tools/install.bzl, line 240 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_27_5jUdpEvTXlJUf:-KueeR8KQtwSBmlWMlGv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L240)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah, I think I can follow it now.  I still think this is a bad-code-smell place to perform the path escaping / quoting.  It is extremely spooky action-at-a-distance (well, really invariants-maintined-at-a-distance).

It's the bash emission where the quoting matters, so it seems to me like the body of `create_java_launcher` is where the quoting should happen.  The classpath should be passed down as a list in all bzl and python (the list would be repr'd into the generated install python), and the implementation of `create_java_launcher` should join and quote it.
</blockquote></details>

Done.

---

*[tools/install.bzl, line 305 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-Ku_1kynL6VEftkzi4gk:-KuejaUOM8zkOgd2iMWP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install.bzl#L305)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Sure, I understand what case we're trying to handle.  What's still unclear is why having failed all of the prior if-tests, we know for sure that we have an executable script from the source directory?  Are there truly no other kinds of targets?

I'm looking for either a comment explaining why only this one kind of target could possibly remain, or replacing the `else` guard with an if-test that positively identifies executable scripts from the source directory.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/third_party/libbot/bot-procman-sheriff, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZTMp_id4Z4t5DrLAf:-KuemmXIYmXV-lhFMw4e:brpydtm) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/third_party/libbot/bot-procman-sheriff#L1)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

It is indeed a reimplementation. Should I move it directly into `tools`? That doesn't seem to be the right place either.
</blockquote></details>

Maybe `tools/install/...` somewhere?  I don't have strong preference, just not in the third_party section.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-Kuemw72vwJungYUWN3l:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuenE1gMzxjoZZcU4Gt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 7 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/install/libbot/bot-procman-sheriff, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-KuZTMp_id4Z4t5DrLAf:-Kuey8xgHzcxmi5tigWO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/20e91f9187820d0e5873fc78e206d64aee824d56/tools/install/libbot/bot-procman-sheriff#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Maybe `tools/install/...` somewhere?  I don't have strong preference, just not in the third_party section.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
+@sherm1 platform review please

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-KuezArmD-lObCrO9gcg:b-xso5tn)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@sherm1 I think it's okay to count me for platform, if you want.

sherm1 commented:
Thanks, @jwnimmer-tri. -@sherm1

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7068#-:-Kuf8G-mPeO02prCWH3K:b-6mno6c)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
This is ready to go, other than the curation tag.  It looks like the commits are possibly already curated, in which case just remove the tag and merge.  Of if you want just one commit, remove the tag and squash-merge.
